internal/wasi/wasi_test.go

```diff
@@ -958,7 +959,168 @@ func createFile(t *testing.T, path string, contents []byte) (*memFile, *MemFS) {
+	store, ctx, fn := instantiateWasmStore(t, FunctionPathOpen, ImportPathOpen, moduleName, func(a *wasiAPI) {
+		// randSouce is used to determine the new fd. Fix it to the expectedFD for testing.
+		a.randSource = func(b []byte) error {
+			binary.LittleEndian.PutUint32(b, expectedFd)
+			return nil
+		}
+		api = a // for later tests
+	})
+
+	// TestSnapshotPreview1_PathOpen uses a matrix because setting up test files is complicated and has to be clean each time.
+	type pathOpenFn func(ctx publicwasm.ModuleContext, fd, lookupFlags, path, pathLen, oFlags uint32,
+		fsRightsBase, fsRightsInheriting uint64,
+		fdFlags, resultFD uint32) wasi.Errno
+	tests := []struct {
+		name     string
+		pathOpen func() pathOpenFn
+	}{
+		{"SnapshotPreview1.PathOpen", func() pathOpenFn {
+			return api.PathOpen
+		}},
+		{FunctionPathOpen, func() pathOpenFn {
+			return func(ctx publicwasm.ModuleContext, fd, lookupFlags, path, pathLen, oFlags uint32,
+				fsRightsBase, fsRightsInheriting uint64,
+				fdFlags, resultFD uint32) wasi.Errno {
+				ret, err := store.Engine.Call(ctx.(*wasm.ModuleContext), fn, uint64(fd), uint64(lookupFlags), uint64(path), uint64(pathLen), uint64(oFlags), uint64(fsRightsBase), uint64(fsRightsInheriting), uint64(fdFlags), uint64(resultFD))
+				require.NoError(t, err)
+				return wasi.Errno(ret[0])
+			}
+		}},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+		t.Run(tc.name, func(t *testing.T) {
+			// Create a memFS for testing that has "./wazero" file.
+			memFS := &MemFS{
+				Files: map[string][]byte{
+					"wazero": []byte(""),
+				},
+			}
+			api.opened = map[uint32]fileEntry{
+				fd: {
+					path:    ".",
+					fileSys: memFS,
+				},
+			}
+
+			maskMemory(store, len(expectedMemory))
+			copy(store.Memories[0].Buffer[0:], initialMemory)
+
+			errno := tc.pathOpen()(ctx, fd, lookupFlags, path, pathLen, oFlags, fsRightsBase, fsRightsInheriting, fdFlags, resultFD)
+			require.Equal(t, wasi.ErrnoSuccess, errno)
+			require.Equal(t, expectedMemory, store.Memories[0].Buffer)
+			require.Equal(t, "wazero", api.opened[expectedFd].path) // verify the file was actually opened
+		})
+	}
+}
+
+func TestSnapshotPreview1_PathOpen_Erros(t *testing.T) {
+	validFD := uint64(3) // arbitrary valid fd after 0, 1, and 2, that are stdin/out/err
+	store, ctx, fn := instantiateWasmStore(t, FunctionPathOpen, ImportPathOpen, moduleName, func(a *wasiAPI) {
+		// Create a memFS for testing that has "./wazero" file.
+		memFS := &MemFS{
+			Files: map[string][]byte{
+				"wazero": []byte(""),
+			},
+		}
+		a.opened = map[uint32]fileEntry{
+			uint32(validFD): {
+				path:    ".",
+				fileSys: memFS,
+			},
+		}
+	})
+
+	tests := []struct {
+		name                                                                                        string
+		fd, lookupFlags, path, pathLen, oFlags, fsRightsBase, fsRightsInheriting, fdFlags, resultFD uint64
+		memory                                                                                      []byte
```

Comment:
I don't think we need to test partial writes to memory unless you are using this to know which position in the source you got to (due to multiple possibility of the same code) if this is why, maybe comment it.

Reply:
wow, thanks for the feedback early! Could you tell me more about the meaning of "test partial writes"? Sorry, maybe I'm not following you.